### ISSUE TITLE
Connection Reset by Peer

### ISSUE THREAD

**Issue by iCHAIT** [mfcovington/pubmed-lookup#3]
Hi, I am trying to fetch abstract of around 1000 papers from Pubmed using the `pmid`.
After fetching abstract of about 5-6 papers I get the following error -

```shell
Traceback (most recent call last):
  File "pubmed_abstract_scraper.py", line 20, in <module>
    publication = Publication(lookup)
  File "/usr/local/lib/python3.7/site-packages/pubmed_lookup/pubmed_lookup.py", line 42, in __init__
    self.set_article_url(resolve_doi=resolve_doi)
  File "/usr/local/lib/python3.7/site-packages/pubmed_lookup/pubmed_lookup.py", line 186, in set_article_url
    response = urlopen(doi_url)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 641, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 563, in error
    result = self._call_chain(*args)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 755, in http_error_302
    return self.parent.open(new, timeout=req.timeout)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 543, in _open
    '_open', req)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 1360, in https_open
    context=self._context, check_hostname=self._check_hostname)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 1320, in do_open
    r = h.getresponse()
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1321, in getresponse
    response.begin()
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 296, in begin
    version, status, reason = self._read_status()
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 257, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/socket.py", line 589, in readinto
    return self._sock.recv_into(b)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 1052, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 911, in read
    return self._sslobj.read(len, buffer)
ConnectionResetError: [Errno 54] Connection reset by peer
```

I understand that this is a TCP/IP error and not possible a bug with this library, but what can I do to resolve this?

I am using Python 3.7.3 on macOS 10.14.4 I installed python using brew.

**Comment by sammosummo** [comment]
Is it always the same publication?

I'm getting a similar (possibly the same) error always on the same paper, that was added to PubMed very recently. All other papers work fine.

**Comment by mfcovington** [comment]
@sammosummo Which paper are you having the issue with?

**Comment by sammosummo** [comment]
https://www.ncbi.nlm.nih.gov/pubmed/?term=31665216

**Comment by mfcovington** [comment]
@sammosummo That one is working fine for me. Are you still having the issue? Also, be sure you aren't using the URL you pasted, since it is in the wrong format.

    $ pubmed-citation https://www.ncbi.nlm.nih.gov/pubmed/31665216
    Writing Committee for the ENIGMA-CNV Working Group ., van der Meer D, Sønderby IE, Kaufmann T, Walters GB, et al. (2019). Association of Copy Number Variation of the 15q11.2 BP1-BP2 Region With Cortical and Subcortical Morphology and Cognition. JAMA Psychiatry 1-11.


    $ pubmed-citation 31665216
    Writing Committee for the ENIGMA-CNV Working Group ., van der Meer D, Sønderby IE, Kaufmann T, Walters GB, et al. (2019). Association of Copy Number Variation of the 15q11.2 BP1-BP2 Region With Cortical and Subcortical Morphology and Cognition. JAMA Psychiatry 1-11.

**Comment by iCHAIT** [comment]
@mfcovington 

I tried the same paper (with the correct URL format) using the following code snippet - 


```python
from pubmed_lookup import PubMedLookup
from pubmed_lookup import Publication

email = 'abc@xyz.com'

url = 'https://www.ncbi.nlm.nih.gov/pubmed/31665216'

lookup = PubMedLookup(url, email)

publication = Publication(lookup) 
```

Strangely, it gives me the error - `[Errno 54] Connection reset by peer`

Maybe there is some difference between  `pubmed-citation` and `PubMedLookup`?

**Comment by mfcovington** [comment]
This is happening because Pubmed returns the DOI URL rather than the URL to the actual publication. When you lookup a publication using this package, the DOI URL gets resolved into the actual URL by default. This particular URL has a pop-up ad and takes a while too load, so may be hitting the global default timeout. You can override resolving the DOI URL with `resolve_doi=False` which solves your problem:

    >>> publication = Publication(lookup, resolve_doi=False) 
    >>> publication.url
    'http://dx.doi.org/10.1001/jamapsychiatry.2019.3779'

When I get a chance, I will go ahead and catch this error and leave the URL as the DOI URL when the error occurs. Or maybe I can make the timeout value customizable.

**Comment by mfcovington** [comment]
FYI: This issue has been addressed in version 0.2.2.

**Comment by sammosummo** [comment]
> @mfcovington
> 
> I tried the same paper (with the correct URL format) using the following code snippet -
> 
> ```python
> from pubmed_lookup import PubMedLookup
> from pubmed_lookup import Publication
> 
> email = 'abc@xyz.com'
> 
> url = 'https://www.ncbi.nlm.nih.gov/pubmed/31665216'
> 
> lookup = PubMedLookup(url, email)
> 
> publication = Publication(lookup) 
> ```
> 
> Strangely, it gives me the error - `[Errno 54] Connection reset by peer`
> 
> Maybe there is some difference between `pubmed-citation` and `PubMedLookup`?

Yep, this is exactly the problem.
